dcpy/connectors/s3.py

```diff
@@ -8,9 +18,87 @@ def client(
     endpoint_url=os.environ["AWS_S3_ENDPOINT"],
 ):
     """Returns a client for AWS S3."""
+    config = Config(read_timeout=120)
     return boto3.client(
         "s3",
         aws_access_key_id=aws_access_key_id,
         aws_secret_access_key=aws_secret_access_key,
         endpoint_url=endpoint_url,
+        config=config,
     )
+
+
+def upload_file(
+    bucket: str,
+    path: str,
+    key: str,
+    acl: str = "public-read",
+    metadata: dict = None,
+) -> dict:
+    """Uploads a single file to AWS S3.
+    adapted from data library
+    https://github.com/NYCPlanning/db-data-library/blob/main/library/s3.py#L37"""
+    with Progress(
+        SpinnerColumn(spinner_name="earth"),
+        TextColumn("[progress.description]{task.description}"),
+        BarColumn(bar_width=30),
+        TextColumn("[progress.percentage]{task.percentage:>3.0f}%"),
+        TimeRemainingColumn(),
+        transient=True,
+    ) as progress:
+        size = os.stat(path).st_size
+        task = progress.add_task(
+            f"[green]Uploading [bold]{os.path.basename(path)}[/bold]", total=size
+        )
+
+        response = client().upload_file(
+            path,
+            bucket,
+            key,
+            ExtraArgs={"ACL": acl, "Metadata": metadata},
+            Callback=lambda complete: progress.update(task, completed=complete),
+        )
+    return response
+
+
+def download_folder(
+    bucket, prefix, export_path, include_prefix_in_export=False
+) -> None:
+    """given bucket, prefix filter, and export path, download contents of folder from s3 recursively"""
```

Comment:
So does this just download a folder and it's non-folder children? Recursively makes me think you'd be recursing to download a full folder tree? (seems fine either way... maybe safer the first way, so you don't accidentally download a full bucket)

Reply:
for later readers, yes, it is recursive because of how prefixes work

Reply:
Chatted in person but for posterity (aka @damonmcc ) - list_objects does find ALL objects with the given path prefix. So downloading a whole bucket (or first 1000, believe is default for list_objects) is certainly possible.